syntax = "proto2";

import "TSPMessages.proto";
package TSK;

enum StructuredTextImportType {
  ImportType_UNKNOWN = 0;
  ImportType_DELIMITED = 1;
  ImportType_FIXEDWIDTH = 2;
}
message TreeNode {
  optional string name = 1;
  repeated .TSP.Reference children = 2;
  optional .TSP.Reference object = 3;
}

message LocalCommandHistoryItem {
  optional .TSP.Reference command = 1;
  optional .TSP.Reference behavior = 2;
}

message LocalCommandHistoryArray {
  required .TSP.LargeArray large_array = 1;
}

message LocalCommandHistoryArraySegment {
  required .TSP.LargeObjectArraySegment large_object_array_segment = 1;
}

message LocalCommandHistory {
  required uint32 undo_count = 1;
  optional .TSP.Reference items_array = 2;
  optional bool fixed_radar_13365177 = 10;
}

message DocumentArchive {
  optional string locale_identifier = 4;
  optional .TSP.Reference annotation_author_storage = 7;
  repeated .TSP.Reference activity_log_entries = 8;
  optional string creation_locale_identifier = 9;
  optional bool prevent_image_conversion_on_open = 10;
  optional bool has_floating_locale = 11;
  optional bool has_user_defined_locale = 12;
  optional .TSP.Reference collaboration_operation_history = 14;
  optional bool should_measure_negatively_tracked_text_correctly = 15;
  optional bool use_optimized_text_vertical_alignment = 16;
  optional bool should_allow_ligatures_in_minimally_tracked_text = 18;
  optional .TSK.FormattingSymbolsArchive formatting_symbols = 17;
  optional .TSP.Reference activity_stream = 199;
}

message FormattingSymbolsArchive {
  message CurrencySymbol {
    required string code = 1;
    required string symbol = 2;
  }

  optional string version = 1;
  optional string calendar = 2;
  optional string numbering_system = 3;
  repeated string months = 4;
  repeated string standalone_months = 5;
  repeated string short_months = 6;
  repeated string standalone_short_months = 7;
  repeated string weekdays = 8;
  repeated string standalone_weekdays = 9;
  repeated string short_weekdays = 10;
  repeated string standalone_short_weekdays = 11;
  optional string am_symbol = 12;
  optional string pm_symbol = 13;
  repeated string tiny_months = 14;
  repeated string standalone_tiny_months = 15;
  repeated string tiny_weekdays = 16;
  repeated string standalone_tiny_weekdays = 17;
  repeated string quarters = 18;
  repeated string standalone_quarters = 19;
  repeated string short_quarters = 20;
  repeated string standalone_short_quarters = 21;
  repeated string eras = 22;
  repeated string long_eras = 23;
  optional string short_date_pattern = 24;
  optional string medium_date_pattern = 25;
  optional string long_date_pattern = 26;
  optional string full_date_pattern = 27;
  optional string short_time_pattern = 28;
  optional string medium_time_pattern = 29;
  optional string long_time_pattern = 30;
  optional string full_time_pattern = 31;
  optional string decimal_separator = 32;
  optional string grouping_separator = 33;
  optional string currency_decimal_separator = 34;
  optional string currency_grouping_separator = 35;
  optional string plus_sign = 36;
  optional string minus_sign = 37;
  optional string exponential_symbol = 38;
  optional string percent_symbol = 39;
  optional string per_mille_symbol = 40;
  optional string infinity_symbol = 41;
  optional string nan_symbol = 42;
  optional string decimal_pattern = 43;
  optional string scientific_pattern = 44;
  optional string percent_pattern = 45;
  optional string currency_pattern = 46;
  optional string currency_code = 47;
  repeated .TSK.FormattingSymbolsArchive.CurrencySymbol currency_symbols = 48;
}

message DocumentSupportArchive {
  optional .TSP.Reference command_history = 1;
  optional uint32 undo_count = 4;
  optional uint32 redo_count = 5;
  optional string undo_action_string = 6;
  optional string redo_action_string = 7;
  optional .TSP.Reference web_state = 8;
  optional bool is_in_collaboration_mode = 9 [default = false];
  optional string action_string_localization = 12;
  optional .TSP.Reference collaboration_state = 13;
  optional .TSP.Reference activity_notification_map = 14;
  optional .TSP.Reference removed_author_auditor_pending_state = 15;
  optional .TSP.Reference command_selection_behavior_history = 2 [deprecated = true];
  optional .TSP.Reference view_state = 3 [deprecated = true];
  optional .TSP.Reference collaboration_command_history = 10 [deprecated = true];
  optional .TSP.Reference collaboration_session_state = 11 [deprecated = true];
}

message ViewStateArchive {
  required .TSP.Reference view_state_root = 1;
  optional int32 document_revision_sequence = 2;
  optional string document_revision_identifier = 3;
}

message CommandArchive {
  optional .TSP.Reference undoRedoState = 1 [deprecated = true];
  optional .TSP.Reference undoCollection = 2;
  optional bool shadowed_by_transform = 3;
  optional bool shadowed_by_commit = 4;
  optional bool remote = 5;
  optional bool should_hold_until_group_commit = 6;
  optional bool server_originated = 7;
}

message CommandGroupArchive {
  required .TSK.CommandArchive super = 1;
  repeated .TSP.Reference commands = 2;
  optional .TSP.IndexSet process_results = 3;
  optional string action_string = 4;
  optional bool can_coalesce_group = 5;
}

message InducedCommandCollectionArchive {
  required .TSK.CommandArchive super = 1;
  optional .TSP.Reference inducing_command = 2;
  repeated .TSP.Reference induced_commands = 3;
  optional .TSP.IndexSet indexes_of_processed_induced_commands = 4;
}

message PropagatedCommandCollectionArchive {
  required .TSK.CommandArchive super = 1;
  required .TSP.Reference propagatable_command = 2;
  optional bool propagatable_command_process_result = 3 [default = false];
  required .TSP.Reference propagating_command = 4;
  optional bool propagating_command_process_result = 5 [default = false];
}

message FinalCommandPairArchive {
  required .TSK.CommandArchive super = 1;
  required .TSP.Reference command = 2;
  optional bool command_process_result = 3 [default = false];
  required .TSP.Reference final_command = 4;
  optional bool final_command_process_result = 5 [default = false];
}

message CommandContainerArchive {
  repeated .TSP.Reference commands = 1;
}

message ProgressiveCommandGroupArchive {
  required .TSK.CommandGroupArchive super = 1;
}

message FormatStructArchive {
  optional uint32 format_type = 1;
  optional uint32 decimal_places = 2;
  optional string currency_code = 3;
  optional uint32 negative_style = 4;
  optional bool show_thousands_separator = 5;
  optional bool use_accounting_style = 6;
  optional uint32 duration_style = 7;
  optional uint32 base = 8;
  optional uint32 base_places = 9;
  optional bool base_use_minus_sign = 10;
  optional uint32 fraction_accuracy = 11;
  optional bool suppress_date_format = 12;
  optional bool suppress_time_format = 13;
  optional string date_time_format = 14;
  optional uint32 duration_unit_largest = 15;
  optional uint32 duration_unit_smallest = 16;
  optional uint32 custom_id = 17;
  optional string custom_format_string = 18;
  optional double scale_factor = 19;
  optional bool requires_fraction_replacement = 20;
  optional double control_minimum = 21;
  optional double control_maximum = 22;
  optional double control_increment = 23;
  optional uint32 control_format_type = 24;
  optional uint32 slider_orientation = 25;
  optional uint32 slider_position = 26;
  optional uint32 decimal_width = 27;
  optional uint32 min_integer_width = 28;
  optional uint32 num_nonspace_integer_digits = 29;
  optional uint32 num_nonspace_decimal_digits = 30;
  optional uint32 index_from_right_last_integer = 31;
  repeated string interstitial_strings = 32;
  optional .TSP.IndexSet inters_str_insertion_indexes = 33;
  optional uint32 num_hash_decimal_digits = 34;
  optional uint32 total_num_decimal_digits = 35;
  optional bool is_complex = 36;
  optional bool contains_integer_token = 37;
  optional uint32 multiple_choice_list_initial_value = 38;
  optional uint32 multiple_choice_list_id = 39;
  optional bool use_automatic_duration_units = 40;
  optional .TSP.UUID custom_uid = 41;
  optional .TSK.CustomFormatArchive custom_format = 42;
  optional bool uses_plus_sign = 43;
  optional string bool_true_string = 44;
  optional string bool_false_string = 45;
  extensions 10000 to 20000;
}

message CustomFormatArchive {
  message Condition {
    required uint32 condition_type = 1;
    optional float condition_value = 2;
    required .TSK.FormatStructArchive condition_format = 3;
    optional double condition_value_dbl = 4;
  }

  required string name = 1;
  required uint32 format_type_pre_bnc = 2;
  required .TSK.FormatStructArchive default_format = 3;
  repeated .TSK.CustomFormatArchive.Condition conditions = 4;
  optional uint32 format_type = 5;
}

message CustomFormatListArchive {
  repeated .TSP.UUID uuids = 1;
  repeated .TSK.CustomFormatArchive custom_formats = 2;
}

message AnnotationAuthorArchive {
  optional string name = 1;
  optional .TSP.Color color = 2;
  optional string public_id = 3;
  optional bool is_public_author = 4;
  repeated string public_ids = 5;
}

message DeprecatedChangeAuthorArchive {
  optional string name = 1;
  optional .TSP.Color change_color = 2;
}

message AnnotationAuthorStorageArchive {
  repeated .TSP.Reference annotation_author = 1;
}

message CommandBehaviorSelectionPathStorageArchive {
  optional .TSK.SelectionPathArchive archived_selection = 1;
  optional .TSK.SelectionPathArchive archived_old_selection = 2;
  optional .TSK.SelectionPathArchive archived_new_selection = 3;
  optional .TSP.Reference forward_selection_path_transformer = 4;
  optional .TSP.Reference reverse_selection_path_transformer = 5;
}

message CommandBehaviorArchive {
  optional .TSP.Reference selection_behavior = 1;
  optional .TSP.Reference activity_behavior = 2;
}

message CommandSelectionBehaviorArchive {
  optional .TSP.Reference selection_path_storage = 1;
  optional uint64 selection_flags = 2;
  optional uint64 additional_forward_selection_flags = 3;
  optional uint64 additional_reverse_selection_flags = 4;
  repeated .TSP.Reference additional_selection_behaviors = 5;
}

message SelectionPathTransformerArchive {
  repeated .TSP.Reference selection_transformers = 1;
}

message SelectionPathArchive {
  repeated .TSP.Reference ordered_selections = 1;
}

message DocumentSelectionArchive {
  optional .TSP.Reference document_root = 1;
}

message NullCommandArchive {
  required .TSK.CommandArchive super = 1;
}

message GroupCommitCommandArchive {
  required .TSK.CommandArchive super = 1;
  optional bool can_coalesce_group = 2;
}

message UpgradeDocPostProcessingCommandArchive {
  required .TSK.CommandArchive super = 1;
}

message InducedCommandCollectionCommitCommandArchive {
  required .TSK.CommandArchive super = 1;
}

message ChangeDocumentPackageTypeCommandArchive {
  enum PackageType {
    Default = 0;
    Directory = 1;
    SingleFile = 2;
  }
  required .TSK.CommandArchive super = 1;
  required .TSK.ChangeDocumentPackageTypeCommandArchive.PackageType new_package_type = 2;
  required .TSK.ChangeDocumentPackageTypeCommandArchive.PackageType old_package_type = 3;
}

message RangeAddress {
  repeated uint64 address_identifier = 1;
  repeated uint32 range_list = 2;
}

message Operation {
  enum OperationType {
    Add = 0;
    Remove = 1;
    Placement = 2;
    Rearrange = 3;
    UpdateId = 4;
    UpdateRange = 5;
    ReplaceRange = 6;
  }
  optional .TSK.Operation.OperationType type = 1 [default = ReplaceRange];
  optional bool noop = 2 [default = false];
  repeated fixed64 address_identifier = 3 [packed = true];
  optional uint64 insert_length = 4 [default = 1];
  optional bool preserve_lower_priority_location = 5 [default = false];
  repeated uint32 range_list = 6 [packed = true];
  optional uint32 transform_behavior = 7 [default = 7];
  optional uint32 property_id = 8;
  optional int32 from_index = 9 [default = -1];
  optional int32 to_index = 10 [default = -1];
  optional bool dominating = 11 [default = false];
  optional int32 object_count = 12 [default = -1];
  optional int32 object_counter_space = 13 [default = 0];
}

message OperationTransformer {
  required bool higher_priority = 1;
  repeated .TSK.Operation operations = 2;
}

message OutgoingCommandQueueItem {
  optional .TSP.Reference command = 1;
  optional string serialized_json_without_data_base64_encoded_string = 2;
  optional .TSP.DataReference serialized_json_without_data_base64_encoded_data = 4;
  optional bool did_rollback_reapply = 6;
  optional bool contains_large_pending_upload_data = 7;
  repeated .TSK.OutgoingCommandQueueItemUUIDToDataMapEntry uuid_to_data_map_entries = 3;
  repeated .TSP.DataReference large_data_list = 5;
}

message OutgoingCommandQueueItemUUIDToDataMapEntry {
  required .TSP.UUID uuid = 1;
  required .TSP.DataReference data = 2;
}

message NativeContentDescription {
  optional string app_name = 1;
  optional string app_version = 2;
  optional string document_id = 3;
  repeated .TSP.Reference drawable_descriptions = 4;
}

message StructuredTextImportSettings {
  required .TSK.StructuredTextImportType type = 1;
  required int32 starting_row = 2;
  repeated string decimal_separators = 3;
  repeated string thousands_separators = 4;
  optional bool transpose_rows_and_columns = 5;
  repeated string delimiters = 6;
  repeated string text_qualifiers = 7;
  optional bool collapse_consecutive = 8;
  optional .TSP.IndexSet column_offsets = 9;
  optional bool automatic_delimiters = 10;
  optional bool automatic_offsets = 11;
  optional uint64 source_encoding = 12;
}

message OperationStorageCommandOperationsEntry {
  optional bool command_identifier_same_as_revision_identifier = 1;
  repeated fixed64 command_identifier = 2 [packed = true];
  repeated .TSK.Operation operations = 3;
  optional bool server_originated = 4;
  optional uint64 coalesced_command_entry_count = 5;
}

message OperationStorageEntry {
  repeated fixed64 document_revision_identifier = 1 [packed = true];
  optional int32 document_revision_sequence_delta = 2;
  repeated .TSK.OperationStorageCommandOperationsEntry command_operation_entries = 3;
  optional double first_entry_creation_time = 4;
  optional int32 creation_time_diff_bucket = 5;
  repeated uint32 file_format_version = 6 [packed = true];
}

message DataReferenceRecord {
  message ContainerUUIDToReferencedDataPair {
    required .TSP.UUID container_uuid = 1;
    required .TSP.DataReference referenced_data = 2;
    required uint32 reference_count = 3;
  }

  repeated .TSK.DataReferenceRecord.ContainerUUIDToReferencedDataPair added_container_uuid_to_referenced_data_pairs = 1;
  repeated .TSK.DataReferenceRecord.ContainerUUIDToReferencedDataPair removed_container_uuid_to_referenced_data_pairs = 2;
  repeated .TSP.DataReference unbounded_referenced_datas = 3;
}

message PencilAnnotationUIState {
  enum PencilAnnotationToolType {
    Pen = 0;
    Highlighter = 1;
  }
  optional .TSK.PencilAnnotationUIState.PencilAnnotationToolType current_tool_type = 1 [default = Pen];
  optional .TSP.Color pen_tool_color = 2;
  optional float pen_tool_opacity = 3;
  optional float pen_tool_width = 4;
  optional .TSP.Color highlighter_tool_color = 5;
  optional float highlighter_tool_opacity = 6;
  optional float highlighter_tool_width = 7;
}
